src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -644,6 +644,88 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
             }
         }
 
+        protected override void ApplyProjectChanges(ProjectChanges projectChanges)
+        {
+            base.ApplyProjectChanges(projectChanges);
+
+            var changedMappedDocuments = projectChanges.GetChangedDocuments()
+                .Where(id => ShouldApplyChangesToMappedDocuments(id))
+                .ToImmutableArray();
+            if (changedMappedDocuments.Any())
+            {
+                ApplyChangesForMappedDocuments(projectChanges, changedMappedDocuments);
+            }
+
+            bool ShouldApplyChangesToMappedDocuments(DocumentId id)
+            {
+                var document = this.CurrentSolution.GetDocument(id);
+                // Only consider files that are mapped and that we are unable to apply changes to.
+                return document?.Services?.GetService<ISpanMappingService>() != null && document?.CanApplyChange() == false;
+            }
+        }
+
+        private void ApplyChangesForMappedDocuments(ProjectChanges projectChanges, ImmutableArray<DocumentId> changedMappedDocuments)
+        {
+            // Get the original text changes from all documents and call the span mapping service to get span mappings for the text changes.
+            var mappedSpansPerDocument = GetTextChangesAndMappedSpans(projectChanges, changedMappedDocuments);
+
+            // Create mapped text changes using the mapped spans and original text changes.
+            using var _ = ArrayBuilder<(string FileName, TextChange TextChange)>.GetInstance(out var mappedChangesBuilder);
+            foreach (var mappedSpans in mappedSpansPerDocument)
+            {
+                var newDoc = projectChanges.NewProject.GetDocument(mappedSpans.Key)!;
+                MapTextChanges(newDoc, mappedSpans.Value.MappedSpans, mappedSpans.Value.OriginalTextChanges, mappedChangesBuilder);
+            }
+
+            var mappedChanges = mappedChangesBuilder.ToImmutable();
+            if (mappedChanges.Any())
+            {
+                // Group the mapped text changes by file, then apply all mapped text changes for the file.
+                foreach (var changesForFile in mappedChanges.GroupBy(change => change.FileName))
+                {
+                    using var invisibleEditor = new InvisibleEditor(ServiceProvider.GlobalProvider, changesForFile.Key, GetHierarchy(projectChanges.ProjectId), needsSave: true, needsUndoDisabled: false);
```

Comment:
We don't have a IServiceProvider object floating around _somewhere_ in VisualStudioWorkspace?

Reply:
All the instances I could find in VSworkspace used the GlobalProvider